### PR TITLE
Add support for merge with memory for aggregation spill

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java [hunk 1]
```diff
@@ -122,7 +124,7 @@ public void tearDown()
     }
 
     @Test(dataProvider = "hashEnabledAndMemoryLimitBeforeSpillValues")
-    public void testHashAggregation(boolean hashEnabled, long memoryLimitBeforeSpill)
+    public void testHashAggregation(boolean hashEnabled, long memoryLimitBeforeSpill, long memoryLimitForMergeWithMemory)
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:127]: Make sure that this tests are actually executing `mergeFromDiskAndMemory` (for example with debugger)

Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:127]: I have checked with debugger that both paths execute

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 2]
```diff
@@ -46,10 +49,13 @@
     private final Optional<Integer> hashChannel;
     private final OperatorContext operatorContext;
     private final long memorySizeBeforeSpill;
+    private final long memoryLimitForMergeWithMemory;
     private Optional<Spiller> spiller = Optional.empty();
     private Optional<MergingHashAggregationBuilder> merger = Optional.empty();
     private CompletableFuture<?> spillInProgress = CompletableFuture.completedFuture(null);
-    private final LocalMemoryContext systemMemoryContext;
+    private final LocalMemoryContext aggregationMemoryContext;
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java:56]: you could make two separate commits (both for better clarity and for better stats on github :) ). One refactoring those contexts and second adding merge with memory

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 3]
```diff
@@ -70,8 +77,13 @@ public SpillableHashAggregationBuilder(
...
         this.operatorContext = operatorContext;
         this.memorySizeBeforeSpill = memoryLimitBeforeSpill.toBytes();
+        this.memoryLimitForMergeWithMemory = memoryLimitForMergeWithMemory.toBytes();
         this.spillerFactory = spillerFactory;
-        this.systemMemoryContext = operatorContext.getSystemMemoryContext().newLocalMemoryContext();
+
+        AbstractAggregatedMemoryContext systemMemoryContext = operatorContext.getSystemMemoryContext();
+        this.aggregationMemoryContext = systemMemoryContext.newLocalMemoryContext();
+        this.spillMemoryContext = systemMemoryContext.newLocalMemoryContext();
+        this.mergeMemoryContext = systemMemoryContext.newLocalMemoryContext();
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: don't make a field but generate it on the fly during construction of `MergingAggregationBuilder`

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java [hunk 4]
```diff
@@ -109,10 +109,12 @@ public void setUp()
     public static Object[][] hashEnabledAndMemoryLimitBeforeSpillValuesProvider()
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:109]: Add dedicated test for merging with memory.

**Comment by pnowojski** [comment]
LGTM % comments

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java:56]: okey

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 3 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java (hunk 1 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:127]: but do they actually process any data?

Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:127]: added test for that

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java (hunk 4 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:109]: done

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java [hunk 5]
```diff
@@ -389,7 +395,44 @@ public void testMultiplePartialFlushes(boolean hashEnabled, long memoryLimitBefo
         }
     }
 
-    private static class DummySpillerFactory implements SpillerFactory
+    @Test
+    public void testMergeWithMemorySpill()
+    {
+        List<Integer> hashChannels = Ints.asList(0);
+        RowPagesBuilder rowPagesBuilder = rowPagesBuilder(BIGINT);
+        List<Page> input = rowPagesBuilder
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:403]: maybe make it

``````
        List<Page> input = rowPagesBuilder
                .addSequencePage(150000, 0)
                .addSequencePage(10, 150000)
                .addSequencePage(10, 150000)
                .build();
```?
``````

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java [hunk 6]
```diff
@@ -389,7 +395,44 @@ public void testMultiplePartialFlushes(boolean hashEnabled, long memoryLimitBefo
...
+                0,
+                new PlanNodeId("test"),
+                ImmutableList.of(BIGINT),
+                hashChannels,
+                Step.SINGLE,
+                ImmutableList.of(LONG_SUM.bind(ImmutableList.of(0), Optional.empty(), Optional.empty(), 1.0)),
+                rowPagesBuilder.getHashChannel(),
+                1,
+                new DataSize(16, MEGABYTE),
+                new DataSize(150, Unit.KILOBYTE),
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:418]: please tie those `150xxx` values to a constant
